### PR TITLE
[Merged by Bors] - feat: `notation3` delaborator generation

### PR DESCRIPTION
**Pull request by kmill**
Gives the `notation3` command the ability to generate a delaborator in most common situations. When it succeeds, `notation3`-defined syntax is pretty printable.

Examples:
```
(⋃ (i : ι) (i' : ι'), s i i') = ⋃ (i' : ι') (i : ι), s i i'
(⨆ (g : α → ℝ≥0∞) (_ : Measurable g) (_ : g ≤ f), ∫⁻ (a : α), g a ∂μ) = ∫⁻ (a : α), f a ∂μ
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kmill** [comment]
@digama0 It's still in CI, but in my local testing it seems to work. It's an extension to the `notation3` command to generate a delaborator, and the output is comparable to what we used to have in Lean 3.

I'm trying to add some usability improvement to mathlib in time for the Lean workshop at MSRI starts next week, so it would be helpful having an expedited review, though worst case we could use a custom branch of mathlib for TPIL.

File: Mathlib/GroupTheory/Perm/Cycle/Concrete.lean [hunk 1]
```diff
@@ -521,8 +521,8 @@ def isoCycle' : { f : Perm α // IsCycle f } ≃ { s : Cycle α // s.Nodup ∧ s
     right_inv := Fintype.leftInverse_bijInv _ }
 #align equiv.perm.iso_cycle' Equiv.Perm.isoCycle'
 
-notation3"c["(l", "* => foldr (h t => List.cons h t) List.nil)"]" =>
-  Cycle.formPerm (↑l) (Cycle.nodup_coe_iff.mpr _)
+notation3 "c["(l", "* => foldr (h t => List.cons h t) List.nil)"]" =>
+  Cycle.formPerm (Cycle.ofList l) (Iff.mpr Cycle.nodup_coe_iff _)
```
[the remark this answers is not in the document]
Comment (digama0) [reply Mathlib/GroupTheory/Perm/Cycle/Concrete.lean:525]: It looks like the syntax here could be reconstructed from the elaborated expression though, although maybe you can't even elaborate this expression in context?

[the remark this answers is not in the document]
Comment (kmill) [reply Mathlib/GroupTheory/Perm/Cycle/Concrete.lean:525]: I'd thought about trying to elaborate the expressions to generate patterns, but then I wasn't sure if there'd be other things that would introduce into the expressions that'd get in the way of being a pattern, like maybe the elaborator might fill in some instance arguments with instances it wouldn't choose for some concrete arguments. With this syntax walker, I felt more confidant that it's doing the right thing in the cases it accepts.

The code's set up where you can swap out `mkExprMatcher` for a fancier implementation

**Comment by kmill** [comment]
In principle there should only be three that don't work right now. `∀ᵐ` and `∃ᵐ` use `<|` in a superficial way, and it's easily fixed either in the declarations or more generally by macro expanding syntax before generating patterns.

The notation for `Lex` in Mathlib.Data.Pi.Lex seems trickier to support, and maybe that one deserves `(parsing_only := true)`.

I'll take a look at them all tomorrow more carefully once I have a compiled mathlib.

**Comment by kmill** [comment]
@digama0 I added a `(prettyPrint := false)` option to `notation3`. It's running through CI, but as far as I understand only one `notation3` needed it, and everything else delaborates.

**Comment by digama0** [comment]
As I discussed with @kmill privately, I have a rewrite of large parts of this code to move more of the matcher algorithm to compile time. But this implementation sets bar for the intended behavior of the delaborator, and I have no complaints about the code itself. So let's get this in the hands of mathlib users.

bors r+

**Comment by ChrisHughes24** [comment]
Conflicts

bors r-

**Pull request closed without merging**